### PR TITLE
Polish GHA configs

### PR DESCRIPTION
**Pull request by Goooler**
- Replace `ubuntu-22.04` with `ubuntu-latest`.
- Remove workarounds for freeing space as the default runner storage has been increased, see https://github.blog/2024-01-17-github-hosted-runners-double-the-power-for-open-source.
- Add Java 21 to LTS test matrix, and use more Java 21 after upgrading to Gradle 8.5
- Remove `groovy-build-test-19.yml`, seems it's unused for now.
- Add more additional test JDKs.
- Remove unnecessary Java matrix.

### CODE REVIEW

**Comment by Goooler** [comment]
JDK 23 EA fails, any idea?

https://github.com/Goooler/groovy/actions/runs/7620906327/job/20756416157

**Comment by daniellansun** [comment]
> JDK 23 EA fails, any idea?
> 
> https://github.com/Goooler/groovy/actions/runs/7620906327/job/20756416157

Don't add JDK 23 build for now.

**Comment by Goooler** [comment]
All checks passed in my fork https://github.com/Goooler/groovy/commit/7c3074a0c7c036a3b8cbad0188ac38de3cc6f6e7.

File: src/test/org/codehaus/groovy/util/ReferenceManagerTest.groovy [hunk 1]
```diff
@@ -144,7 +144,7 @@ class ReferenceManagerTest extends GroovyTestCase {
         }
     }
 
-    private class TestReference<T>
+    private final class TestReference<T>
```
Comment (Goooler) [inline src/test/org/codehaus/groovy/util/ReferenceManagerTest.groovy:147]: May depend on #2037?

Comment (daniellansun) [reply src/test/org/codehaus/groovy/util/ReferenceManagerTest.groovy:147]: Yep

**Comment by Goooler** [comment]
I'll squash these commits when we are ready.

**Review by paulk-asert — APPROVED** [review]
Please squash before committing. If some individual commits are good candidates for cherry-picking to 4_0_X (while others not), they could be a separate commit.

**Comment by Goooler** [comment]
I just merged test jobs into one file in 7f53ec4f4623188b3d5f49a782e0e8df8aabef18, is this makes sense to you?

**Pull request merged**